### PR TITLE
[PR] Adding auth

### LINKED ISSUE #37: Auth!
# todo
+ [x] Create new file: `auth.md`
+ [x] Add tutorial using `auth_plug` to add authentication to the Todo App
+ [x] "Lift and Shift" from: https://github.com/dwyl/phoenix-chat-example/pull/120
+ [x] Tests.

### PR DESCRIPTION
**Pull request by LuchoTurtle**
This PR adds authentication which allows people to filter for `items` they created.
Closes:
+ [x] #37 
+ [x] #63 
+ [x] #65

### CODE REVIEW

**Comment by LuchoTurtle** [comment]
This should be done.
All the test pass and authentication seems to be working properly.

File: README.md [hunk 1]
```diff
@@ -2378,7 +2378,18 @@ Seriously, try the Heroku demo again:
...
-<br />
+### 12 (Bonus!) Adding authentication
+
+Currently, the application allows *anyone* 
+to access it and manage todo items.
+However, wouldn't it be awesome if 
+we added *authentication* so each user
+could check their own list?
+
+We created a [small guide](./auth.md)
```
Comment (nelsonic) [inline README.md:2389]: "small guide" is not a very informative link text. see: https://github.com/dwyl/contributing/issues/119
We're better off with just the file name:
```suggestion
We created a dedicated authentication guide: 
[`/auth.md`](./auth.md)
```

File: lib/app/todo/item.ex [hunk 2]
```diff
@@ -3,7 +3,7 @@ defmodule App.Todo.Item do
   import Ecto.Changeset
 
   schema "items" do
-    field :person_id, :integer, default: 0
+    field :person_email, :string, default: ""
```
Comment (nelsonic) [inline lib/app/todo/item.ex:6]: I don't understand _why_ this is necessary ... 🤷‍♂️

**Review by nelsonic — CHANGES_REQUESTED** [review]
@LuchoTurtle the "Todo" in #37 was to "lift-and-shift" from: https://github.com/dwyl/phoenix-chat-example/pull/120 ...
That added **_optional_ `auth`**. i.e. people could still try the app without authenticating. 
Adding `|> validate_required([:person_email, :text])` to the `item` changeset means that it's no longer ***optional***.  unless you are capturing their `email` address. But then _why_ are we complicating this? 🤷‍♂️

**Comment by LuchoTurtle** [comment]
@nelsonic I knew that an `id` was returned from `auth_plug` but I still decided to use `person_email` because I thought it would be more eligible for beginners. It would also be easier for these to see which "emails" referred to the todo list (instead of having an `id`).
I was aware that it would "complicate a bit" but I did it in good faith so it would be easier for people to see which todos referred to whom.

**Comment by nelsonic** [comment]
We don't want to store email addresses in a random app tutorial app [or any other non-auth app for that matter].
if we deploy this to **`Fly.io`** without encrypting the email addresses then we have to _think_ about how to keep those email addresses [highly personally identifiable information] private.  or we're breaking the [GDPR]() law. 
see: https://www.linkedin.com/pulse/gdpr-you-collect-email-addresses-personal-data-dr-adrian-ioannou/

A `person_id` from `authdemo` is not personally identifiable because it's a system generated `id` 
that gives a hacker no other info about the person and cannot be used maliciously without hacking our `auth` app. 

This isn't a question of "good faith" ... 
it's a question of doing something that wasn't/isn't required 
and ultimately makes the app a target for data theft and us liable.
It's not something we would do in _practice_ so we shouldn't do it in in a tutorial
where a beginner might get the idea that it's "OK" to store email addresses like this.

As much as I want to just ignore this, and get back to my work, I cannot. 
We cannot write tutorials that ignore best practice [and the Law] for data privacy/security.
Emails are appropriately encrypted in the `auth` App:
https://github.com/dwyl/auth/blob/bc1f8a9ca08ff26a4f7a50ef2e71bae4985d33d1/lib/auth/person.ex#L14-L15

Our **`MVP`** uses the `person_id` and has no other personally identifiable information.
This is the appropriate [Separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns).

Please make your own life a _lot_ simpler and try not over-think it. 🙏

**Comment by LuchoTurtle** [comment]
The `auth` version was added as a bonus and was meant to be run on localhost, never wanted it to actually be deployed unencrypted online. Perhaps I didn't make that clear.

I'll change it.

**Comment by nelsonic** [comment]
auth will be deployed to fly.io the same as in https://phoenix-chat.fly.dev otherwise it's not a good showcase.

**Comment by LuchoTurtle** [comment]
@nelsonic just made the changes required.
If you want to review for merge, go ahead.

File: lib/app/todo/item.ex (hunk 2 above)
Comment (LuchoTurtle) [reply lib/app/todo/item.ex]: Fixed with https://github.com/dwyl/phoenix-todo-list-tutorial/pull/56/commits/c9cb6313fe3fbc0e1afde401b3473155fced9613

File: auth.md [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+Change `lib/app/todo/item.ex`
+so it looks like the following.
+
+```elixir
+defmodule App.Todo.Item do
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  schema "items" do
+    field :person_email, :string, default: ""
```
[the remark this answers is not in the document]
Comment (LuchoTurtle) [reply auth.md]: Fixed with https://github.com/dwyl/phoenix-todo-list-tutorial/pull/56/commits/c9cb6313fe3fbc0e1afde401b3473155fced9613

**Comment by nelsonic** [comment]
@LuchoTurtle should the tests be passing for this PR? 
https://github.com/dwyl/phoenix-todo-list-tutorial/actions/runs/3661224877/jobs/6189330825#step:8:37
image

**Comment by LuchoTurtle** [comment]
[The tests are failing (tossing that [warning line 1 on router](warning: this clause cannot match because a previous clause at line 1 always matches) warning you mentioned before) for some odd reason - it works on a blank slate on my laptop and it was working prior to just fixing the merge conflicts 🤷](https://github.com/dwyl/phoenix-todo-list-tutorial/pull/58#issuecomment-1344859468)

**Comment by LuchoTurtle** [comment]
I will take a closer look tomorrow 👍

**Comment by LuchoTurtle** [comment]
The CI should be fixed by now 👍

**Comment by LuchoTurtle** [comment]
Added a few tests so coverage increases enough for CI to run.

**Comment by nelsonic** [comment]
```
git checkout 'auth'
git pull
mix deps.get
mix test

ompiling 10 files (.ex)
Generated app app
......warning: variable "item" is unused (if the variable is not meant to be used, prefix it with an underscore)
  test/app_web/controllers/item_controller_test.exs:85: AppWeb.ItemControllerTest."test clear completed clears the completed items"/1

warning: variable "items" is unused (if the variable is not meant to be used, prefix it with an underscore)
  test/app_web/controllers/item_controller_test.exs:91: AppWeb.ItemControllerTest."test clear completed clears the completed items"/1

warning: variable "item" is unused (if the variable is not meant to be used, prefix it with an underscore)
  test/app_web/controllers/item_controller_test.exs:98: AppWeb.ItemControllerTest."test clear completed clears the completed items in public (person_id=0)"/1

warning: variable "items" is unused (if the variable is not meant to be used, prefix it with an underscore)
  test/app_web/controllers/item_controller_test.exs:104: AppWeb.ItemControllerTest."test clear completed clears the completed items in public (person_id=0)"/1

warning: key :givenName will be overridden in map
  test/app_web/controllers/item_controller_test.exs:149

....09:26:46.686 [error] ERROR: AUTH_API_KEY Environment Variable is not set


  1) test get /logout with valid JWT (AppWeb.AuthControllerTest)
     test/app_web/controllers/auth_controller_test.exs:21
     ** (FunctionClauseError) no function clause matching in Regex.run/3

     The following arguments were given to Regex.run/3:

         # 1
         ~r/^(.*)\/(.*)\/(.*)$/

         # 2
         nil

         # 3
         []

     Attempted function clauses (showing 1 out of 1):

         def run(%Regex{} = regex, string, options) when is_binary(string)

     code: jwt = AuthPlug.Token.generate_jwt!(data)
     stacktrace:
       (elixir 1.14.1) lib/regex.ex:335: Regex.run/3
       (auth_plug 1.5.1) lib/token.ex:52: AuthPlug.Token.client_secret/0
       (auth_plug 1.5.1) lib/token.ex:86: AuthPlug.Token.generate_jwt!/1
       test/app_web/controllers/auth_controller_test.exs:31: (test)

warning: module attribute @completed_attrs was set but never used
  test/app_web/controllers/item_controller_test.exs:9

09:26:46.843 request_id=FzJ2Otal0L_oqbcAAHqH [error] ERROR: AUTH_API_KEY Environment Variable is not set
09:26:46.849 [error] ERROR: AUTH_API_KEY Environment Variable is not set

... more similar failing tests ....

  9) test update item redirects when data is valid (AppWeb.ItemControllerTest)
     test/app_web/controllers/item_controller_test.exs:70
     ** (FunctionClauseError) no function clause matching in Regex.run/3

     The following arguments were given to Regex.run/3:

         # 1
         ~r/^(.*)\/(.*)\/(.*)$/

         # 2
         nil

         # 3
         []

     Attempted function clauses (showing 1 out of 1):

         def run(%Regex{} = regex, string, options) when is_binary(string)

     code: conn = setup_conn(conn)
     stacktrace:
       (elixir 1.14.1) lib/regex.ex:335: Regex.run/3
       (auth_plug 1.5.1) lib/token.ex:52: AuthPlug.Token.client_secret/0
       (auth_plug 1.5.1) lib/token.ex:86: AuthPlug.Token.generate_jwt!/1
       test/app_web/controllers/item_controller_test.exs:173: AppWeb.ItemControllerTest.setup_conn/1
       test/app_web/controllers/item_controller_test.exs:71: (test)

........
Finished in 0.3 seconds (0.2s async, 0.1s sync)
34 tests, 9 failures

Randomized with seed 570274
```

My reading of the `README.md` in it's current state is that there is no reference to `AUTH_API_KEY` Environment Variable:
[dwyl/phoenix-todo-list-tutorial/blob/auth/README.md](https://github.com/dwyl/phoenix-todo-list-tutorial/blob/4461617a388092b8d9071710b2682af2f0390b20/README.md)

This will impact the _next_ person's ability to run the _finished_ project on their `localhost` ... 💔

**Comment by nelsonic** [comment]
Created `.env_sample` file with `AUTH_API_KEY` so that tests _always_ pass on `localhost`: https://github.com/dwyl/phoenix-todo-list-tutorial/pull/56/commits/4238996e7a628f3de40e61a764fc8e4337f1d57f


Now tidying up these warnings:

```sh
mix test
warning: variable "item" is unused (if the variable is not meant to be used, prefix it with an underscore)
  test/app_web/controllers/item_controller_test.exs:85: AppWeb.ItemControllerTest."test clear completed clears the completed items"/1

warning: variable "items" is unused (if the variable is not meant to be used, prefix it with an underscore)
  test/app_web/controllers/item_controller_test.exs:91: AppWeb.ItemControllerTest."test clear completed clears the completed items"/1

warning: variable "item" is unused (if the variable is not meant to be used, prefix it with an underscore)
  test/app_web/controllers/item_controller_test.exs:98: AppWeb.ItemControllerTest."test clear completed clears the completed items in public (person_id=0)"/1

warning: variable "items" is unused (if the variable is not meant to be used, prefix it with an underscore)
  test/app_web/controllers/item_controller_test.exs:104: AppWeb.ItemControllerTest."test clear completed clears the completed items in public (person_id=0)"/1

warning: key :givenName will be overridden in map
  test/app_web/controllers/item_controller_test.exs:149

..........warning: module attribute @completed_attrs was set but never used
  test/app_web/controllers/item_controller_test.exs:9

........................
Finished in 0.7 seconds (0.6s async, 0.1s sync)
34 tests, 0 failures

Randomized with seed 76640
```

**Comment by nelsonic** [comment]
OK no more warnings:
```sh
mix c
..................................
Finished in 0.6 seconds (0.5s async, 0.07s sync)
34 tests, 0 failures

Randomized with seed 910810
----------------
COV    FILE                                        LINES RELEVANT   MISSED
100.0% lib/app/todo.ex                               112        7        0
100.0% lib/app/todo/item.ex                           19        2        0
100.0% lib/app_web/controllers/auth_controller.       14        2        0
100.0% lib/app_web/controllers/error_html.ex          19        1        0
100.0% lib/app_web/controllers/error_json.ex          15        1        0
 97.8% lib/app_web/controllers/item_controller.      123       45        1
100.0% lib/app_web/controllers/item_html.ex           41       13        0
  0.0% lib/app_web/gettext.ex                         24        0        0
100.0% lib/app_web/router.ex                          35        8        0
[TOTAL]  98.7%
----------------
Generating report...
```

But still missing that crucial last line of code coverage to reach **`100%`** ... 💭 
Going to add the missing test now. 🧑‍💻

**Comment by LuchoTurtle** [comment]
> But still missing that crucial last line of code coverage to reach 100% ... 💭
This fix might entail tinkering with the endpoints at `router.ex` and changing a few tests. I had to deal with this because of #65 when working on https://github.com/dwyl/phoenix-papertrail-demo/pull/3

**Comment by nelsonic** [comment]
Final thing is to update the GIF in the `README.md` ... 

phx-todo-list-example

## "_It's the little details that are vital. <br />Little things make big things happen._" <br />~ John Wooden

**Comment by nelsonic** [comment]
Then I started re-reading `auth.md` and see code blocks like this:

```elixir
case Map.has_key?(conn.assigns, :person) do
  false ->
    items = Todo.list_items()
    changeset = Todo.change_item(item)

    render(conn, "index.html",
      items: items,
      changeset: changeset,
      editing: item,
      filter: Map.get(params, "filter", "all")
    )

  true ->
    person_id = Map.get(conn.assigns.person, :id)

    items = Todo.list_items(person_id)
    changeset =

    render(conn, "index.html",
      items: Todo.list_items(person_id),
      changeset: Todo.change_item(item),
      editing: item,
      filter: Map.get(params, "filter", "all")
    )
end
```

[DRY](https://en.wikipedia.org/wiki/Don't_repeat_yourself)ed to:

```elixir
render(conn, "index.html",
  items: Todo.list_items(Useful.get_in_default(conn, [:assigns, :person, :id], 0)),
  changeset: Todo.change_item(item),
  editing: item,
  filter: Map.get(params, "filter", "all")
)
```

**Review by nelsonic — APPROVED** [review]
@LuchoTurtle great work. Thanks. 🎉

**Pull request merged**